Refactor `mouse_entered` and `mouse_exited` signals

Sauermann opened the pull request:
fix variant of #20881 where Controls are in different Viewports
fix #66293
fix no longer #57349 (that is already fixed for the case of a single Viewport)
prerequisite for #67531
prerequisite for #77781
addresses #48634 after #77933
related to #54565 (maybe supersedes it)
done: don't merge before #67903 (fix display-server differences between Windows and linuxbsd)

The current implementation for signals `mouse_entered` and `mouse_exited` have shortcomings that relate to focused windows and pressed mouse buttons. For example a `Control` can be hovered by mouse, even if it is occluded by an embedded `Window`.

This patch changes the behavior, so that `Control` and `Viewports` send their mouse-enter/exit-notifications based solely on mouse position, visible area and input restrictions and not on which window has focus or which mouse buttons are pressed. This implicitly also changes when the `mouse_entered` and `mouse_exited` signals are sent.

Rewriting a core functionality like `mouse_entered` and `mouse_exited` should be tested thoroughly. While I tested it in lots of situations and had a look at the locations where it is used in the code base, it can be, that I have missed something.

So far, tests on the following platforms were successful:
- linuxbsd
- Windows
- Web (Firefox, Chromium, Edge)

Other platforms:
- macOS: not yet tested
- Android / iOS: Not sure, if the concept of `mouse_entered` and `mouse_exited` makes sense in this context.
- Android: Signals do not trigger with touch and drag

The concept for this implementation is:
- When a mouse-event is sent from the DisplayServer to a Window, it first recursively determines mouse-over states for all Viewports and sends mouse-enter/exit notifications and signals.
- Afterwards the event handling happens just as before, without the enter/exit-notifications.
- Additional adjustments to how mouse-over-state is dropped were necessary.

MRP: [BugVisualMouseNotifications.zip](https://github.com/godotengine/godot/files/9865023/BugVisualMouseNotifications.zip) (updated 2022-10-26)

This functionality can not be implemented as a part of `Viewport::_gui_input_event`, because of its interplay with Windows and because `Viewport::_gui_input_event` is based on input and not on visibility.

https://user-images.githubusercontent.com/6299227/197390115-33c8b18d-69ff-4d87-a293-8378fe099db9.mp4

Updated 2022-11-13: Fix location of `sw->_update_mouse_over` call.
Updated 2022-11-15: Fix merge conflict with #68019
Updated 2022-11-19: Fix location of `sw->_update_mouse_over` call.
Updated 2023-01-06: Fix merge conflict with #69972
Updated 2023-01-27: Fix merge conflict with #66692
Updated 2023-02-18: Fix merge conflict with #72764
Updated 2023-06-07: Fix merge conflict with #77933
Updated 2023-06-29: Made necessary adjustments for Gui in 3D Demo
Updated 2023-07-07: Fix merge conflict with #78078
Updated 2023-07-27: Fix merge conflict with #79248

lmurray commented:
I disagree with adding new signals that differ from the existing ones as a single mouse cursor can only be in one location at a time so it cannot be both inside and outside of a region at the same time which is what is implied by having multiple signals. No other UI toolkit that I know has two separate entered/exited signals that have different rules. Instead I think the existing signals are incorrectly implemented and need to be modified to correctly detect when the cursor enters and exits a region under all circumstances.

Sauermann commented:
@lmurray I also would prefer to replace the old signals. However i have not yet tested the places, where the old signals are used within the code-base. So I provided new signals in this initial version of this PR, so that they can be tested and compared.

bruvzg reviewed (commented):
Adding a new signals/notifications is probably unnecessary, it should replace existing viewport signals/methods `NOTIFICATION_VP_MOUSE_*` (for native windows there's `NOTIFICATION_WM_MOUSE_*`).

Sauermann commented:
I have updated the PR, so that now `Viewports` emit `NOTIFICATION_VP_MOUSE_*` notifications based on the new method.
The behavior changes, but I didn't experience any problematic effects.

I left `NOTIFICATION_WM_MOUSE_*` alone, because this works already in the expected way.

My next step will be investigating, if replacing the `Control`-notifications by the new method would introduce unwanted behavior.

Sauermann commented:
With the latest update, I have implemented a replacement for the `Control`-signals `mouse_enter/exit`. Feedback about how this works on other platforms than Linux would be appreciated.

Sauermann commented:
Since the Windows and linuxbsd Display server behave quite differently, it makes sense to merge this PR only after the differences are solved. The PR #67903 makes this happen.

Sauermann committed:
Refactor mouse_entered and mouse_exited notifications

The previous implementation for signals mouse_entered and mouse_exited
had shortcomings that relate to focused windows and pressed mouse buttons.
For example a Control can be hovered by mouse, even if it is occluded by
an embedded window.

This patch changes the behavior, so that Control and Viewport send
their mouse-enter/exit-notifications based solely on mouse position,
visible area, and input restrictions and not on which window has
focus or which mouse buttons are pressed. This implicitly also
changes when the mouse_entered and mouse_exited signals are sent.

This functionality can not be implemented as a part of
Viewport::_gui_input_event, because of its interplay with Windows and
because Viewport::_gui_input_event is based on input and not on
visibility.

KoBeWi commented:
I gave it a test and found a bug very quickly:
![mycha_yfOC70IBoB](https://user-images.githubusercontent.com/2223172/212781042-9a5f79f7-e395-4ef8-95c3-e03821a7e1c4.gif)
The cursor shape misbehaves when going from connection dialog to script editor. Without this PR it consistently uses caret shape.

Sauermann commented:
Thanks for the feedback @KoBeWi 
This issue was caused by the Windows Display Server which did not send a `WINDOW_EVENT_MOUSE_EXIT` event in the case that a window got closed.

KoBeWi commented on doc/classes/Node.xml:
```diff
@@ -927,10 +927,10 @@
 		</constant>
 		<constant name="NOTIFICATION_VP_MOUSE_ENTER" value="1010">
-			Notification received when the mouse enters the viewport.
+			Notification received when the mouse cursor enters the [Viewport]'s visible area, that is not concealed behind other [Control]s or [Window]s, provided its [member Viewport.gui_disable_input] is [code]false[/code] and regardless if it's currently focused or not.
```
Can a Control obscure a Viewport?
btw you seem to use `[]` inconsistently compared to descriptions above.

Sauermann replied:
With the latest update, I use [] consistently in the docs.

bruvzg commented:
Seems to break popup `y` positions on macOS (but it might be unrelated and just need a rebase, I'll retest later). Couldn't find any other bugs, mouse enter/exit seems to work ok.

<img width="404" alt="Screenshot 2023-01-26 at 13 13 14" src="https://user-images.githubusercontent.com/7645683/214824887-82bd03ca-61d5-4582-af82-5471ba505472.png">

Sauermann commented:
I'm not sure, if this should be moved to 4.1.
It has potential for regressions.
@bruvzg have you had any chance to retest, if your problem was actually caused by this PR or just required a rebase?

Inve1951 commented:
I don't quite understand the issue regarding the `mouse_entered`/`mouse_exited` signals. They fire correctly on 4.0.2 when dragging a child control outside its parent. But maybe I misunderstand - is it about firing them on the child itself?

What does not work in 4.0.2 however is the *mouse_up* event. It fires at the control where the user initiated the drag operation, and flags the event as handled, supressing both `_unhandled_input` and `_gui_input` at the cursor location - effectively making drag&drop behaviors impossible.
**Does this PR address that?**

See https://github.com/godotengine/godot/issues/20881#issuecomment-867027437 and https://github.com/godotengine/godot/issues/20881#issuecomment-883817011.

#54565 adds the boolean flag `Viewport.gui_force_mouse_events` which supposedly fixes that without a breaking change (while I personally think it should be the new default). But https://github.com/godotengine/godot/pull/54565#issuecomment-1427029304 suggests that #54565 is superseded by this PR (#67791).

Sauermann commented:
@Inve1951 

> I don't quite understand the issue regarding the `mouse_entered`/`mouse_exited` signals. They fire correctly on 4.0.2 when dragging a child control outside its parent. But maybe I misunderstand - is it about firing them on the child itself?

Indeed, the issue is that a Control previously didn't receive `mouse_entered`/`mouse_exited`, while a Control-node is dragged. 

> What does not work in 4.0.2 however is the _mouse_up_ event. It fires at the control where the user initiated the drag operation, and flags the event as handled, supressing both `_unhandled_input` and `_gui_input` at the cursor location - effectively making drag&drop behaviors impossible. **Does this PR address that?**

No, this PR doesn't take care of that issue. However I aim to solve that issue with #67531.

> #54565 adds the boolean flag `Viewport.gui_force_mouse_events` which supposedly fixes that without a breaking change (while I personally think it should be the new default). But [#54565 (comment)](https://github.com/godotengine/godot/pull/54565#issuecomment-1427029304) suggests that #54565 is superseded by this PR (#67791).

If I remember correctly, #54565 wouldn't solve the problem of missing `mouse_entered`/`mouse_exited` events in other Windows / SubViewports.

Sauermann commented:
> Seems to break popup `y` positions on macOS (but it might be unrelated and just need a rebase, I'll retest later). Couldn't find any other bugs, mouse enter/exit seems to work ok.

@bruvzg may I ask you to perform the retest on macOS? I am not aware of any Node-positions changes by this PR.

Sauermann commented:
Please move to 4.2

KoBeWi commented:
#57349 is not fixed by this PR.

Sauermann commented:
As we discussed yesterday in #20881, mouse-entered/exited notifications now work (within a single Viewport).

#57349 states as use-case: "So I want to have a generic `is_over_ui` boolean to find out if my mouse is over UI or not."
This use-case can now be accomplished in GDscript by setting `is_over_ui` based on mouse-entered/exited notifications.

So I agree, that this PR doesn't solve the use-case of #57349 (in the scenario of a single viewport), because it already works. I will update the OP accordingly.

KoBeWi reviewed (approved):
I reviewed the code and it seems fine (not like I know much about it though) and the behavior is correct.
The issue I mentioned should be unlinked, unless you manage to fix it.

Sauermann commented:
Thanks for the review. I have unlinked the mentioned issue.

YuriSizov commented:
@RandomShaper Could you take a look here?

RandomShaper commented on doc/classes/Control.xml:
```diff
@@ -1140,10 +1138,10 @@
 		</constant>
 		<constant name="NOTIFICATION_MOUSE_ENTER" value="41">
-			Sent when the mouse pointer enters the node.
+			Sent when the mouse cursor enters the control's visible area, that is not concealed behind other Controls or Windows, provided its [member mouse_filter] lets the event reach it and regardless if it's currently focused or not.
```
I have to admit I need to use the Eng-Spa dictionary to understand 'concealed.' Could you use a moer common word?

lmurray replied:
"Occluded" is the de-facto technical term for this.

kleonc replied:
Is this statement true though? AFAICT after skimming through the code the sorting based on which these mouse notifcations occur doesn't take z-index into account. Which would mean it's not always a matter of the actual visible areas. :thinking:
https://github.com/godotengine/godot/blob/c3769e46b8a82640e14fbc8fca029787af757536/scene/main/viewport.cpp#L1623-L1693
https://github.com/godotengine/godot/blob/c3769e46b8a82640e14fbc8fca029787af757536/scene/main/viewport.cpp#L1381-L1389
https://github.com/godotengine/godot/blob/da81ca62a5f6d615516929896caa0b6b09ceccfc/scene/gui/control.h#L152-L160
https://github.com/godotengine/godot/blob/da81ca62a5f6d615516929896caa0b6b09ceccfc/scene/main/node.cpp#L1759-L1813

Sauermann replied:
Thank you for pointing out the z-index. I did miss that.

Now I have been trying to describe the behavior in a way that includes all details but coming up with a concise way to do that is really hard. I hope, the current version is acceptable. I will gladly accept suggestions how to describe this in a better way.

RandomShaper commented on scene/main/viewport.cpp:
```diff
@@ -2949,6 +2930,156 @@ bool Viewport::_sub_windows_forward_input(const Ref<InputEvent> &p_event) {
+		return;
+	}
+
+	if (get_tree()->get_root()->is_embedding_subwindows() || Object::cast_to<SubViewport>(this)) {
```
This reminds me of something we discussed in another PR, that it felt awkard to check the type of `this` that way. However, I don't remember where that was nor the decision. Do you know what I'm talking about?

Sauermann replied:
Yes I remember, that was https://github.com/godotengine/godot/pull/77923#discussion_r1222519549
I will look into making this less awkward.

Sauermann replied:
I have implemented this in a different way.

RandomShaper reviewed (commented):
What can I say? This is extremely complex. As others have said, code looks OK. Also, this has been reasonably tested, so I'd give it greenlight with a big TIWAGOS to let it be more exhaustively tested at this early point of the release cycle. I don't think there's a better way of proceeding unless we want to risk this rotting and we losing the benefits that it can provide.

RandomShaper commented on scene/main/window.cpp:
```diff
@@ -677,16 +677,24 @@ void Window::_event_callback(DisplayServer::WindowEvent p_event) {
 			_propagate_window_notification(this, NOTIFICATION_WM_MOUSE_ENTER);
-			emit_signal(SNAME("mouse_entered"));
+			Window *root = get_tree()->get_root();
+#ifdef DEBUG_ENABLED
+			ERR_FAIL_COND_MSG(root->gui.windowmanager_window_over, "Entering a window while a window is hovered should never happen.");
+#endif // DEBUG_ENABLED
```
If that should never happen, as in "this is an engine bug," maybe a `DEV_ASSERT` is better (remember the condition has to be reversed for that macro).

Sauermann replied:
Indeed this was intended for engine bugs. I agree, that `DEV_ASSERT` is a better fit.

RandomShaper reviewed (approved):
Let's put this in the wild!

lawnjelly commented:
I wrote this on rocketchat earlier, as I was not sure whether `DEV_ASSERT` should have been used in this case (I originally wrote the macro, and Juan did warn of this potential problem at the time):

> Sauermann Just to note, I don't know whether DEV_ASSERT was needed there originally versus just printing an error and doing nothing (i.e. would it crash immediately afterwards)?
> In general we should prefer things like ERR_FAIL wherever possible because we don't want the engine to crash when there are bugs, and DEV_ASSERT will not act outside DEV_ENABLED builds.
> It should also be for performance critical stuff, there's no reason why there shouldn't be error checking in release builds here I think.
> I'll make a PR making note of this.

1xch commented:
I just updated to v4.2.dev3.official [013e8e3af] from v4.1.1 and I'm having a problem with what I suspect is this in particular.
![Untitled](https://github.com/godotengine/godot/assets/199137/8a17c83c-36bd-41db-a42e-cd5380589341)
With the image above, the left is the window, the right is a snapshot of the node structure.  G_Window & G_Pane have widgets in them for handling capture, that previously just listened to mouse enter & exit. For example:

```
func _on_ready():
	_home.mouse_entered.connect(func(): LXN_FOCUSED.emit(true))
	_home.mouse_exited.connect(func(): LXN_FOCUSED.emit(false))
``` 

Now they do not work at all. Can I get a quick glance at this and maybe someone say why it needs to be this way when it did work. I'd like to know if this might be a bug, and I need to provide more info or documentation or if I need to rewrite how I'm handling capture on this. I'd really rather not as it was more a pain to poll mouse events before I just settled on using mouse enter and exit as the main part of measuring capture.

Sauermann commented:
@1xch from your short description, it is difficult to guess, what the problem might be.
Also I'm not entirely sure, what you mean by "capture". In your code you emit a signal called "LXN_FOCUSED". Is the problem with mouse-capture or with focus?
If you could provide a minimal reproduction project of the problem, that you are experiencing, I could have a deeper look at it.

I just noticed, that you use a `SubViewport`, that is not a child of a `SubViewportContainer`. Have you implemented a way for sending InputEvents to that `SubViewport`?

1xch commented:
@Sauermann,
1. What I am saying is that the above worked. I've had it working for a while, with continual tinkering through most 4.0 versions, the above has been working for most of this year, with recent refactors this summer.
2. The problem is with mouse_entered & mouse_exited signals is what I've traced: they no longer happen. I would mostly ignore my terminology, I posted the above after discovery. 'Capture' here means acquiring mouse or other activity indicating the Control needs focused.
3. Yes, I've implemented a way for InputEvents to go to the SubViewport. That is not the issue.

I think a clearer statement is that G_Window & G_Pane in the above are controls that previously responded with mouse_entered & mouse_exited signals and now they do not. Both nodes still receive InputEvents, my core issue being the signals and not knowing why they aren't triggering when

Sauermann commented:
@1xch Thanks for the MRP. I have tested it in v4.0.stable.official [92bee43ad], v4.1.1.stable.official [bd6af8e0e] and in v4.2.dev.custom_build [fa3428ff2] on Linux.

Your MRP behaved identical in my tests: when moving the mouse from `G_Window` to `G_Pane`, `G_Window` did receive a mouse-exited signal. So I can't replicate the overlapping behavior, that you describe. Please verify, if the MRP works for you as you would expect in v4.1.1.

I have tested the behavior in 3.5.1 and also there the parent Control receives a mouse-exited signal, when the mouse moves to a child-Control.

I would also kindly ask you to make future MRP less complicated: You use multiple `instantiantiate()` calls and different source-code files, which makes the project difficult and time-consuming to understand.

wouldntsavezion commented:
Has there been a discussion about this somewhere? I see the work started back in 2022. I personally don't understand why the `mouse_entered` and `mouse_exited` signals behavior was changed. Also, this seems to be the source of many issues now, as well as effectively making `MOUSE_FILTER_PASS` largely unusable. 

Props to @Sauermann for all the work and now adding PRs to notify users in the docs, but this seems to me like it's just letting users know that events will not behave correctly. I think more discussion should definitely happen in #82530 , which sums up the situation quite well.

emirljuca commented:
Yeah I gotta say, this broke my game too, I was relying on mouse pass to pass the mouse signal down while still being able to use the mouse entered and mouse exited signal. I have a control node that is further down the node stack that uses the mouse signal for a tooltip that will no longer work. Why make these changes? Is there any plans to add mouse pass work?

kitbdev commented:
@emirljuca That issue should already be resolved in master as of #84547.

emirljuca commented:
> @emirljuca That issue should already be resolved in master as of #84547.

That is great to hear, testing on newest build seems to show this to be the case, thanks!
